src/jdk.jlink/share/classes/jdk/tools/jlink/internal/runtimelink/ResourceDiff.java

```diff
@@ -0,0 +1,288 @@
+ * which the optimized image got derived from (e.g. packaged modules). The
+ * differences are being used in JRTArchive so as to back-track from an optimized
+ * jimage to the original (i.e. it restores original resources using the diff).
+ */
+public class ResourceDiff implements Comparable<ResourceDiff> {
+
+    private static final int MAGIC = 0xabba;
+
+    public static enum Kind {
+        ADDED((short)1),    // Resource added
+        REMOVED((short)2),  // Resource removed
+        MODIFIED((short)3); // Resource modified
+
+        private short value;
+
+        private Kind(short value) {
+            this.value = value;
+        }
+
+        public short value() {
+            return value;
+        }
+
+        static Kind fromShort(short v) {
+            if (v > 3 || v < 1) {
+                throw new IllegalArgumentException("Must be within range [1-3]");
+            }
+            switch (v) {
+            case 1: return ADDED;
+            case 2: return REMOVED;
+            case 3: return MODIFIED;
+            }
+            throw new IllegalStateException("Must not reach here!");
+        }
+    }
+
+    private final Kind kind;
+    private final byte[] resourceBytes;
+    private final String name;
+
+    private ResourceDiff(Kind kind, String name, byte[] resourceBytes) {
+        this.kind = kind;
+        this.name = name;
+        if ((kind == Kind.REMOVED || kind == Kind.MODIFIED) &&
+                resourceBytes == null) {
+            throw new AssertionError("Resource bytes must be set for REMOVED or MODIFIED");
+        }
+        this.resourceBytes = resourceBytes;
+    }
+
+    public Kind getKind() {
+        return kind;
+    }
+
+    public byte[] getResourceBytes() {
+        return resourceBytes;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public int compareTo(ResourceDiff o) {
+        int kindComp = kind.value() - o.kind.value();
+        if (kindComp == 0) {
+            return getName().compareTo(o.getName());
+        } else {
+            return kindComp;
+        }
+    }
+
+    public static class Builder {
+        private Kind kind;
+        private String name;
+        private byte[] resourceBytes;
+
+        public Builder setKind(Kind kind) {
+            this.kind = kind;
+            return this;
+        }
+        public Builder setName(String name) {
+            this.name = Objects.requireNonNull(name);
+            return this;
+        }
+        public Builder setResourceBytes(byte[] resourceBytes) {
+            this.resourceBytes = Objects.requireNonNull(resourceBytes);
+            return this;
+        }
+        public ResourceDiff build() {
+            if (kind == null || name == null) {
+                // should this be RuntimeImageLinkException?   output needs cleanup?
+                throw new IllegalStateException("kind and name must be set");
```

Comment:
It's my leftover comment.   Will this happen?  Seems not - then should be AssertionError.

Same question in line 144-145